### PR TITLE
Extract type parameter names

### PR DESCRIPTION
**Pull request by lachmatt**
## Why

Include type parameter names in stacktraces.

## What

Cloned and modified `GetSigTypeTokName` from `clr_helpers.cpp` to extract type parameter names.

## Tests

Existing tests for AlwaysOnProfiling.

### CODE REVIEW

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 1]
```diff
@@ -337,17 +343,36 @@ class SamplingHelper
...
+        HCORENUM functionGenParamsIter = nullptr;
+        HCORENUM classGenParamsIter = nullptr;
+        mdGenericParam functionGenericParams[generic_params_max_len]{};
+        mdGenericParam classGenericParams[generic_params_max_len]{};
+        ULONG functionGenParamsCount = 0;
+        ULONG classGenParamsCount = 0;
+
+        mdTypeDef classToken = function_info.type.id;
+
+        pIMDImport->EnumGenericParams(&classGenParamsIter, classToken, classGenericParams, generic_params_max_len,
```
Comment (Kielek) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:356]: consider `hr = pIMDImport`
and then  if (FAILED(hr))

both for method and the class

Consider the same behaviour as for failing functionMethodSignature.TryParse

**Review by Kielek — APPROVED** [review]

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp (hunk 1 above)
Comment (lachmatt) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp]: fixed

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 2]
```diff
@@ -356,12 +395,167 @@ class SamplingHelper
...
+        ULONG num = 0;
+        pbCur += CorSigUncompressData(pbCur, &num);
+        if (num >= generic_params_max_len)
+        {
+            return unknown;
+        }
+        WCHAR param_type_name[param_name_max_len]{};
+        ULONG pch_name = 0;
+        const auto hr = pImport->GetGenericParamProps(genericParameters[num], nullptr, nullptr, nullptr, nullptr,
+                                                      param_type_name, param_name_max_len - 1, &pch_name);
```
Comment (pjanotti) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:418]: not from this PR itself but we need a consistent way to treat calls that break over multiple lines. I don't think we are being consistent about them.

Comment (pjanotti) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:418]: `param_name_max_len - 1`: doesn't the API already handles the space for the terminating `\0`?

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 3]
```diff
@@ -356,12 +395,167 @@ class SamplingHelper
...
+            {
+                pbCur++;
+                tokenName = GetSigTypeTokName(pbCur, pImport, classParams, methodParams) + WStr("[]");
+                break;
+            }
+            case ELEMENT_TYPE_GENERICINST:
+            {
+                pbCur++;
+                tokenName = GetSigTypeTokName(pbCur, pImport, classParams, methodParams);
+                tokenName += WStr("[");
```
Comment (pjanotti) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:524]: Isn't more typical to use angle brackets here?

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 4]
```diff
@@ -337,17 +343,50 @@ class SamplingHelper
...
+        mdGenericParam functionGenericParams[generic_params_max_len]{};
+        mdGenericParam classGenericParams[generic_params_max_len]{};
+        ULONG functionGenParamsCount = 0;
+        ULONG classGenParamsCount = 0;
+
+        mdTypeDef classToken = function_info.type.id;
+
+        hr = pIMDImport->EnumGenericParams(&classGenParamsIter, classToken, classGenericParams, generic_params_max_len,
+                                     &classGenParamsCount);
+        pIMDImport->CloseEnum(classGenParamsIter);
```
[the remark this answers is not in the document]
Comment (lachmatt) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:358]: yes

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 5]
```diff
@@ -337,17 +343,50 @@ class SamplingHelper
...
+            Logger::Debug("Method generic parameters enumeration failed. HRESULT=0x", std::setfill('0'), std::setw(8),
+                          std::hex, hr);
+            result.append(unknown_list_of_arguments);
+            return;
+        }
+        
         // try to list arguments type
-        FunctionMethodSignature function_method_signature = function_info.method_signature;
-        hr = function_method_signature.TryParse();
+        FunctionMethodSignature functionMethodSignature = function_info.method_signature;
```
[the remark this answers is not in the document]
Comment (lachmatt) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:380]: Our naming is inconsistent. I tried to make it more consistent, but decided to postpone it.

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp (hunk 3 above)
Comment (lachmatt) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:524]: This was copied from `clr_helpers.cpp` as is, but I can adjust it.

Comment (Kielek) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:524]: After internal discussions, we will keep square brackets. It it aligned with stacktraces.

**Pull request merged**